sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamComplexTypeTest.java

```diff
@@ -412,32 +398,141 @@ public void testNullDatetimeFields() {
+                        + " EXTRACT(DAY from nullableDateTypeField) as day_with_null, "
+                        + " dateTypeField + interval '1' day as date_with_day_added, "
+                        + " nullableDateTypeField + interval '1' day as day_added_with_null "
+                        + " from PCOLLECTION"));
+
+    Schema outputRowSchema =
+        Schema.builder()
+            .addField("dd", FieldType.INT64)
+            .addNullableField("day_with_null", FieldType.INT64)
+            .addField("date_with_day_added", FieldType.logicalType(SqlTypes.DATE))
+            .addNullableField("day_added_with_null", FieldType.logicalType(SqlTypes.DATE))
+            .build();
+
+    PAssert.that(outputRow)
+        .containsInAnyOrder(
+            Row.withSchema(outputRowSchema)
+                .addValues(27L, null, LocalDate.of(2019, 6, 28), null)
+                .build());
+
+    pipeline.run().waitUntilFinish(Duration.standardMinutes(2));
+  }
+
+  @Test
+  public void testNullSqlLogicalTypeTimeFields() {
+    Schema dateTimeFieldSchema =
+        Schema.builder()
+            .addField("timeTypeField", FieldType.logicalType(SqlTypes.TIME))
+            .addNullableField("nullableTimeTypeField", FieldType.logicalType(SqlTypes.TIME))
+            .build();
+
+    Row timeRow =
+        Row.withSchema(dateTimeFieldSchema).addValues(LocalTime.of(1, 0, 0), null).build();
+
+    PCollection<Row> outputRow =
+        pipeline
+            .apply(Create.of(timeRow))
+            .setRowSchema(dateTimeFieldSchema)
+            .apply(
+                SqlTransform.query(
+                    "select timeTypeField + interval '1' hour as time_with_hour_added, "
+                        + " nullableTimeTypeField + interval '1' hour as hour_added_with_null, "
+                        + " timeTypeField - INTERVAL '60' SECOND as time_with_seconds_added, "
+                        + " nullableTimeTypeField - INTERVAL '60' SECOND as seconds_added_with_null "
+                        + " from PCOLLECTION"));
+
+    Schema outputRowSchema =
+        Schema.builder()
             .addField("time_with_hour_added", FieldType.logicalType(SqlTypes.TIME))
             .addNullableField("hour_added_with_null", FieldType.logicalType(SqlTypes.TIME))
             .addField("time_with_seconds_added", FieldType.logicalType(SqlTypes.TIME))
             .addNullableField("seconds_added_with_null", FieldType.logicalType(SqlTypes.TIME))
+            .build();
+
+    PAssert.that(outputRow)
+        .containsInAnyOrder(
+            Row.withSchema(outputRowSchema)
+                .addValues(LocalTime.of(2, 0, 0), null, LocalTime.of(0, 59, 0), null)
+                .build());
+
+    pipeline.run().waitUntilFinish(Duration.standardMinutes(2));
+  }
+
+  @Test
+  public void testSqlLogicalTypeDatetimeFields() {
+    Schema dateTimeFieldSchema =
+        Schema.builder()
+            .addField("dateTimeField", FieldType.logicalType(SqlTypes.DATETIME))
```

Comment:
Add some tests on "nullableDateTimeField" as well, like above?

Reply:
I planned to do that but Beam do not have `NULLABLE_TIMESTAMP_WITH_LOCAL_TZ` type corresponding to `TIMESTAMP_WITH_LOCAL_TZ`.

Reply:
I don't think we need a `NULLABLE_TIMESTAMP_WITH_LOCAL_TZ` in Calcite. Like in the test above, Calcite does not define a `NULLABLE_DATE` either.

If we add `.addNullableField("nullableDateTimeField", FieldType.logicalType(SqlTypes.DATETIME))`, does that not work?

Reply:
No. I think its because `NULLABLE_DATE` could be recognized but `NULLABLE_TIMESTAMP_WITH_LOCAL_TZ` could not. (https://github.com/apache/beam/blob/5e0e798ddd827fd212ac89b8c6f6f2cf9e4b29a5/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/utils/CalciteUtils.java#L115-L125)

Reply:
Actually is beam type `FieldType.logicalType(sqlTypes.DATETIME).withNullable(true)` could not be recognized.

Reply:
Yeah it makes sense to be in here as well. Thanks for catching this.

Reply:
`NULLABLE_TIMESTAMP_WITH_LOCAL_TZ` and corresponding tests added. It works now.